build.rs

```diff
@@ -1,20 +1,146 @@
+extern crate bindgen;
+
 use std::env;
+use std::path::PathBuf;
 
-fn find_library_paths() -> Vec<String> {
+fn read_env() -> Vec<String> {
     match env::var("CUDA_LIBRARY_PATH") {
         Ok(path) => {
-            let split_char = if cfg!(target_os = "windows") { ";" } else { ":" };
+            let split_char = if cfg!(target_os = "windows") {
+                ";"
+            } else {
+                ":"
+            };
 
-            path.split(split_char).map(|s| s.to_owned()).collect::<Vec<_>>()
+            path.split(split_char)
+                .map(|s| s.to_owned())
+                .collect::<Vec<_>>()
         }
         Err(_) => vec![],
     }
 }
 
-fn main() {
-    for p in find_library_paths() {
-        println!("cargo:rustc-link-search=native={}", p);
+fn find_cuda() -> PathBuf {
+    let mut candidates = read_env();
+    candidates.push("/usr/local/cuda".to_string());
+    candidates.push("/opt/cuda".to_string());
+    for base in &candidates {
+        let base = PathBuf::from(base);
+        let path = base.join("include/cuda.h");
+        if path.is_file() {
+            return base;
+        }
     }
+    panic!("CUDA cannot find");
+}
+
+fn main() {
+    let out_path = PathBuf::from(env::var("OUT_DIR").unwrap());
+    let cuda_path = find_cuda();
+
+    bindgen::builder()
+        .header("wrapper/cuda.h")
+        .clang_arg(format!("-I{}/include", cuda_path.display()))
+        .whitelist_recursively(false)
+        .whitelist_type("^CU.*")
+        .whitelist_type("^cuuint(32|64)_t")
+        .whitelist_type("^cudaError_enum")
+        .whitelist_type("^cudaMem.*")
+        .whitelist_var("^CU.*")
+        .whitelist_function("^CU.*")
+        .whitelist_function("^cu.*")
+        .default_enum_style(bindgen::EnumVariation::Rust)
+        .generate()
+        .expect("Unable to generate CUDA bindings")
+        .write_to_file(out_path.join("cuda_bindings.rs"))
+        .expect("Unable to write CUDA bindings");
+
+    bindgen::builder()
+        .header("wrapper/cublas.h")
+        .clang_arg(format!("-I{}/include", cuda_path.display()))
+        .whitelist_recursively(false)
+        .whitelist_type("^cublas.*")
+        .whitelist_var("^cublas.*")
+        .whitelist_function("^cublas.*")
+        .default_enum_style(bindgen::EnumVariation::Rust)
+        .generate()
+        .expect("Unable to generate CUBLAS bindings")
+        .write_to_file(out_path.join("cublas_bindings.rs"))
+        .expect("Unable to write CUBLAS bindings");
+
+    bindgen::builder()
+        .header("wrapper/cucomplex.h")
+        .clang_arg(format!("-I{}/include", cuda_path.display()))
+        .whitelist_recursively(false)
+        .whitelist_type("^cu.*Complex$")
+        .default_enum_style(bindgen::EnumVariation::Rust)
+        .generate()
+        .expect("Unable to generate CUComplex bindings")
+        .write_to_file(out_path.join("cucomplex_bindings.rs"))
+        .expect("Unable to write CUComplex bindings");
```

Comment:
It seems like this generates bindings for CUComplex and CUBLAS, but only links to CUBLAS. Was that intended? It seems like it should either link to CUComplex as well, or it should not generate bindings for it.

I'd prefer to have separate `cudart-sys`, `cublas-sys` and `cucomplex-sys` crates, and have `cuda-sys` only include the driver API, if that's possible.

Reply:
Looks only 
```rust
pub type cuFloatComplex = float2;
pub type cuDoubleComplex = double2;
pub type cuComplex = cuFloatComplex;
```
is generated from CUComplex.h

Reply:
@bheisler Thanks for the review!

I can add one more problem to the list. cublas doesn't compile using CUDA 8. It's missing the __half and __half2 types. Unfortunately, I couldn't get Bindgen to detect those types in cuda_fp16.h. As a work-around, I've added them manually in https://github.com/LutzCle/accel/commit/74a82a4f856e410ac66f2da90ccdd4e0684c3095.

Reply:
I've create an issue to track *-sys crate separation #9